# Working with .dat files using a PDF containing positional formatting information. 

####  First, I will read in the PDF file that I have found the area for using the Preview App. I have only included the first three columns as that is all I will need today. Column name, beginning position and end posistion. 


In [1]:
from PyPDF2 import PdfFileReader, PdfFileWriter,PdfFileMerger
import tabula

In [2]:
pdf_file = "assign_wk3/NISPUF14_CODEBOOK.PDF"

In [3]:
posdata = tabula.read_pdf(pdf_file, pages="5-21", area=(100.05, 12.34, 767.73, 204))

I tweaked the area so that I could grab all 16 pages at once. The columns almost line up but they are wider on the first page than the last. I changed the area until only 3 complete columns were captured. I checked my work below. 

In [4]:
posdata

[          SEQNUMC    1    6
 0        SEQNUMHH    7   11
 1            PDAT   12   12
 2        PROVWT_D   13   31
 3   PROVWT_D_TERR   32   50
 4         RDDWT_D   51   69
 5    RDDWT_D_TERR   70   88
 6         STRATUM   89   92
 7            YEAR   93   96
 8        AGECPOXR   97   97
 9        HAD_CPOX   98   99
 10       SHOTCARD  100  100
 11         AGEGRP  101  101
 12      BF_ENDR06  102  109
 13     BF_EXCLR06  110  117
 14     BF_FORMR08  118  125
 15      BFENDFL06  126  126
 16     BFFORMFL06  127  127
 17            C1R  128  128
 18            C5R  129  130
 19         CBF_01  131  132
 20        CEN_REG  133  133
 21        CHILDNM  134  134
 22        CWIC_01  135  136
 23        CWIC_02  137  138
 24          EDUC1  139  139
 25        FRSTBRN  140  140
 26       I_HISP_K  141  141
 27       INCPORAR  142  157
 28        INCPOV1  158  158
 29       INCQ298A  159  160
 30          INTRP  161  161
 31       LANGUAGE  162  162
 32       M_AGEGRP  163  163,
             

Above, I have loaded all of the positional data for the dat file. I have uploaded it as one continuous table that I will convert to a csv for coding. 

In [5]:
tabula.convert_into(pdf_file,"data_wk3/posdata_codebook.csv", pages="5-21", area=(100.05, 12.34, 767.73, 204), output_format="csv")

In [6]:
!head data_wk3/posdata_codebook.csv

I save the file to csv, a more easily usable format, and created headers that reflect the original PDF. I confirm no information was lost. 

In [7]:
import csv

In [8]:
with open('code_posdata.csv', 'w', newline='') as out_f:
    writer = csv.writer(out_f)

    with open('data_wk3/posdata_codebook.csv', newline='') as in_f:
        reader = csv.reader(in_f)

        # Read the first row
        first_row = next(reader)
        # Count the columns in first row; equivalent to your `for i in range(len(first_row)): ...`
        header = ['column', 'start', 'stop']

        # Write header and first row
        writer.writerow(header)
        writer.writerow(first_row)

        # Write rest of rows
        for row in reader:
            writer.writerow(row)
            
            # code adapted from https://stackoverflow.com/questions/70544772/how-to-add-a-header-to-an-existing-csv-file-without-replacing-the-first-row

In [9]:
import pandas as pd

In [10]:
df = pd.read_csv('code_posdata.csv')

In [11]:
df.tail(30)

column  start  stop
431  XPOLTY5    820   821
432  XPOLTY6    822   823
433  XPOLTY7    824   824
434  XPOLTY8    825   825
435  XPOLTY9    826   826
436  XROTTY1    827   828
437  XROTTY2    829   830
438  XROTTY3    831   832
439  XROTTY4    833   834
440  XROTTY5    835   836
441  XROTTY6    837   837
442  XROTTY7    838   838
443  XROTTY8    839   839
444  XROTTY9    840   840
445  XVRCTY1    841   842
446  XVRCTY2    843   844
447  XVRCTY3    845   846
448  XVRCTY4    847   847
449  XVRCTY5    848   848
450  XVRCTY6    849   849
451  XVRCTY7    850   850
452  XVRCTY8    851   851
453  XVRCTY9    852   852
454    INS_1    853   854
455    INS_2    855   856
456    INS_3    857   858
457   INS_3A    859   860
458  INS_4_5    861   862
459    INS_6    863   864
460   INS_11    865   866

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461 entries, 0 to 460
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   column  461 non-null    object
 1   start   461 non-null    int64 
 2   stop    461 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 10.9+ KB


So it looks like I have 461 column names in the df above. I will use the df to create a code that will iterate over the dat file using the positions provided in the start and stop columns. 

Below is the data that I will be making useful with the positional data. 

In [13]:
! head assign_wk3/nispuf14.dat

The code below is throwing an error because there are multiple items in the stop column that contain floats and it can't slice using that, so I will remove the .1 from the stop column by rounding them down. 

In [14]:
df1 = df.round()
df1[34:]

column  start  stop
34            MARITAL2    164   164
35             MOBIL_I    165   165
36           NUM_PHONE    166   167
37        NUM_CELLS_HH    168   169
38   NUM_CELLS_PARENTS    170   171
..                 ...    ...   ...
456              INS_3    857   858
457             INS_3A    859   860
458            INS_4_5    861   862
459              INS_6    863   864
460             INS_11    865   866

[427 rows x 3 columns]

In [15]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461 entries, 0 to 460
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   column  461 non-null    object
 1   start   461 non-null    int64 
 2   stop    461 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 10.9+ KB


The code below was developed to read the file line by line if the length of the line is over 1 character. Then store it in an empty dictionary. Then for each row in my dataframe created above, I grab the key from the column named column, start postion from the start column (-1 for python's 0 index), stop position from the stop column, dictionary value is captured using the start and stop values. The empty dictionary key:value. Then the ListofDictionaries is filled with the individual dictionaries created. 

In [19]:
listofdict = []

file = 'assign_wk3/nispuf14.dat'


with open(file) as infile:
    for line in infile:
        if len(line) > 1:
            datdict = {}    #first part of the code adapted from lab week 3 
                
        for i, row in df1.iterrows(): # iterate through each row in df for the following values
            key = row.column
            startdat = int(row.start) - 1
            stopdat = int(row.stop)
            value = line[startdat:stopdat].strip()
            datdict[key] = value
    listofdict.append(datdict)


I check to see that the code worked by pulling the beginning of the list of dictionaries created.  

In [25]:
listofdict[:1]

[{'SEQNUMC': '238811',
  'SEQNUMHH': '23881',
  'PDAT': '1',
  'PROVWT_D': '27.16589416587590',
  'PROVWT_D_TERR': '27.16589416587590',
  'RDDWT_D': '11.64888255983980',
  'RDDWT_D_TERR': '11.64888255983980',
  'STRATUM': '1012',
  'YEAR': '2014',
  'AGECPOXR': '.',
  'HAD_CPOX': '2',
  'SHOTCARD': '2',
  'AGEGRP': '2',
  'BF_ENDR06': '669.6250',
  'BF_EXCLR06': '152.1875',
  'BF_FORMR08': '888.0000',
  'BFENDFL06': '.',
  'BFFORMFL06': '.',
  'C1R': '4',
  'C5R': '2',
  'CBF_01': '1',
  'CEN_REG': '3',
  'CHILDNM': '2',
  'CWIC_01': '2',
  'CWIC_02': '.',
  'EDUC1': '4',
  'FRSTBRN': '1',
  'I_HISP_K': '2',
  'INCPORAR': '3.00000000000000',
  'INCPOV1': '1',
  'INCQ298A': '14',
  'INTRP': '2',
  'LANGUAGE': '1',
  'M_AGEGRP': '3',
  'MARITAL2': '1',
  'MOBIL_I': '2',
  'NUM_PHONE': '1',
  'NUM_CELLS_HH': '2',
  'NUM_CELLS_PARENTS': '2',
  'RACE_K': '1',
  'RACEETHK': '2',
  'RENT_OWN': '2',
  'SEX': '1',
  'ESTIAP14': '12',
  'EST_GRANT': '12',
  'STATE': '11',
  'D6R': '2',
  'D7': '

## I'll save the list of dictionaries created above to a csv for later use. 

with open() closes the file automatically after finishing the block of code, so no need for me to close manually.

In [21]:
myheaders = df.column
myvalues = listofdict

filename = 'outputdict.csv'
with open(filename, 'w', newline='') as myfile:
    writer = csv.DictWriter(myfile, fieldnames=myheaders)
    writer.writeheader()
    writer.writerows(myvalues)
    
    # code adapted from https://www.pythonforbeginners.com/basics/list-of-dictionaries-to-csv-in-python